src/datasets/builder.py

```diff
@@ -1331,63 +1416,183 @@ def _prepare_split(
         split_generator: SplitGenerator,
         check_duplicate_keys: bool,
         file_format="arrow",
+        num_proc: Optional[int] = None,
         max_shard_size: Optional[Union[int, str]] = None,
     ):
+
+        max_shard_size = convert_file_size_to_int(max_shard_size or config.MAX_SHARD_SIZE)
         is_local = not is_remote_filesystem(self._fs)
         path_join = os.path.join if is_local else posixpath.join
 
-        if file_format == "arrow":
-            if max_shard_size is not None:
-                raise NotImplementedError(
-                    "Writing sharded arrow files is not supported. Please don't use max_shard_size or use file_format='parquet'."
-                )
-        else:
-            max_shard_size = convert_file_size_to_int(max_shard_size or config.MAX_SHARD_SIZE)
-
         if self.info.splits is not None:
             split_info = self.info.splits[split_generator.name]
         else:
             split_info = split_generator.split_info
 
-        suffix = "-SSSSS-of-NNNNN" if file_format == "parquet" else ""
-        fname = f"{self.name}-{split_generator.name}{suffix}.{file_format}"
+        SUFFIX = "-JJJJJ-SSSSS-of-NNNNN"
+        fname = f"{self.name}-{split_generator.name}{SUFFIX}.{file_format}"
         fpath = path_join(self._output_dir, fname)
 
-        generator = self._generate_examples(**split_generator.gen_kwargs)
+        # Default to using 16-way parallelism for preparation if the number of files is higher than 16.
+        num_input_shards = _number_of_shards(split_generator.gen_kwargs)
+        if num_input_shards <= 1 and num_proc is not None:
+            logger.warning(
+                f"Setting num_proc from {num_proc} back to 1 for the {split_info.name} split to disable multiprocessing as it only contains one shard."
+            )
+            num_proc = 1
+        elif num_proc is not None and num_input_shards < num_proc:
+            logger.info(
+                f"Setting num_proc from {num_proc} to {num_input_shards} for the {split_info.name} split as it only contains {num_input_shards} shards."
+            )
+            num_proc = num_input_shards
+
+        pbar = logging.tqdm(
+            unit=" examples",
+            total=split_info.num_examples,
+            leave=False,
+            desc=f"Generating {split_info.name} split",
+        )
```

Comment:
"disable" for consistency:

```suggestion
        pbar = logging.tqdm(
            disable=not logging.is_progress_bar_enabled(),
            unit=" examples",
            total=split_info.num_examples,
            leave=False,
            desc=f"Generating {split_info.name} split",
        )
```